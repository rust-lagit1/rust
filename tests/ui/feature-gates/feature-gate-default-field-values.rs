#![feature(generic_const_exprs)]
#![allow(unused_variables, dead_code, incomplete_features)]

pub struct S;

#[derive(Default)]
pub struct Foo {
    pub bar: S = S, //~ ERROR default values on `struct` fields aren't supported
    pub baz: i32 = 42 + 3, //~ ERROR default values on `struct` fields aren't supported
}

#[derive(Default)]
pub enum Bar {
    #[default]
    Foo { //~ ERROR the `#[default]` attribute may only be used on unit enum variants
        bar: S = S, //~ ERROR default values on `struct` fields aren't supported
        baz: i32 = 42 + 3, //~ ERROR default values on `struct` fields aren't supported
    }
}

#[derive(Default)]
pub struct Qux<A, const C: i32> {
    bar: S = Qux::<A, C>::S, //~ ERROR default values on `struct` fields aren't supported
    baz: i32 = foo(), //~ ERROR default values on `struct` fields aren't supported
    bat: i32 = <Qux<A, C> as T>::K, //~ ERROR default values on `struct` fields aren't supported
    bay: i32 = C, //~ ERROR default values on `struct` fields aren't supported
    bak: Vec<A> = Vec::new(), //~ ERROR default values on `struct` fields aren't supported
}

impl<A, const C: i32> Qux<A, C> {
    const S: S = S;
}

trait T {
    const K: i32;
}

impl<A, const C: i32> T for Qux<A, C> {
    const K: i32 = 2;
}

const fn foo() -> i32 {
    42
}

fn main () {
    let x = Foo { .. }; //~ ERROR base expression required after `..`
    let y = Foo::default();
    let z = Foo { baz: 1, .. }; //~ ERROR base expression required after `..`

    assert_eq!(45, x.baz);
    assert_eq!(45, y.baz);
    assert_eq!(1, z.baz);

    let x = Bar::Foo { .. }; //~ ERROR base expression required after `..`
    //~^ ERROR functional record update syntax requires a struct
    let y = Bar::default();
    let z = Bar::Foo { baz: 1, .. }; //~ ERROR base expression required after `..`
    //~^ ERROR functional record update syntax requires a struct

    assert!(matches!(Bar::Foo { bar: S, baz: 45 }, x));
    assert!(matches!(Bar::Foo { bar: S, baz: 45 }, y));
    assert!(matches!(Bar::Foo { bar: S, baz: 1 }, z));

    let x = Qux::<i32, 4> { .. }; //~ ERROR base expression required after `..`
    assert!(matches!(Qux::<i32, 4> { bar: S, baz: 42, bat: 2, bay: 4, .. }, x));
    //~^ ERROR base expression required after `..`
    assert!(x.bak.is_empty());
}
