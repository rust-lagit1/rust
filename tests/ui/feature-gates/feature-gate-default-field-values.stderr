error: the `#[default]` attribute may only be used on unit enum variants
  --> $DIR/feature-gate-default-field-values.rs:15:5
   |
LL |     Foo {
   |     ^^^
   |
   = help: consider a manual implementation of `Default`

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:8:15
   |
LL |     pub bar: S = S,
   |               ^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:9:17
   |
LL |     pub baz: i32 = 42 + 3,
   |                 ^^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:16:15
   |
LL |         bar: S = S,
   |               ^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:17:17
   |
LL |         baz: i32 = 42 + 3,
   |                 ^^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:23:11
   |
LL |     bar: S = Qux::<A, C>::S,
   |           ^^^^^^^^^^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:24:13
   |
LL |     baz: i32 = foo(),
   |             ^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:25:13
   |
LL |     bat: i32 = <Qux<A, C> as T>::K,
   |             ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:26:13
   |
LL |     bay: i32 = C,
   |             ^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: default values on `struct` fields aren't supported
  --> $DIR/feature-gate-default-field-values.rs:27:16
   |
LL |     bak: Vec<A> = Vec::new(),
   |                ^^^^^^^^^^^^^
   |
   = note: see issue #3681 <https://github.com/rust-lang/rust/issues/3681> for more information
   = help: add `#![feature(default_field_values)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:47:21
   |
LL |     let x = Foo { .. };
   |                     ^
   |
help: add a base expression here
   |
LL |     let x = Foo { ../* expr */ };
   |                     ++++++++++

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:49:29
   |
LL |     let z = Foo { baz: 1, .. };
   |                             ^
   |
help: add a base expression here
   |
LL |     let z = Foo { baz: 1, ../* expr */ };
   |                             ++++++++++

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:55:26
   |
LL |     let x = Bar::Foo { .. };
   |                          ^
   |
help: add a base expression here
   |
LL |     let x = Bar::Foo { ../* expr */ };
   |                          ++++++++++

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:58:34
   |
LL |     let z = Bar::Foo { baz: 1, .. };
   |                                  ^
   |
help: add a base expression here
   |
LL |     let z = Bar::Foo { baz: 1, ../* expr */ };
   |                                  ++++++++++

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:65:31
   |
LL |     let x = Qux::<i32, 4> { .. };
   |                               ^
   |
help: add a base expression here
   |
LL |     let x = Qux::<i32, 4> { ../* expr */ };
   |                               ++++++++++

error[E0797]: base expression required after `..`
  --> $DIR/feature-gate-default-field-values.rs:66:73
   |
LL |     assert!(matches!(Qux::<i32, 4> { bar: S, baz: 42, bat: 2, bay: 4, .. }, x));
   |                                                                         ^
   |
help: add a base expression here
   |
LL |     assert!(matches!(Qux::<i32, 4> { bar: S, baz: 42, bat: 2, bay: 4, ../* expr */ }, x));
   |                                                                         ++++++++++

error[E0436]: functional record update syntax requires a struct
  --> $DIR/feature-gate-default-field-values.rs:55:26
   |
LL |     let x = Bar::Foo { .. };
   |                          ^

error[E0436]: functional record update syntax requires a struct
  --> $DIR/feature-gate-default-field-values.rs:58:34
   |
LL |     let z = Bar::Foo { baz: 1, .. };
   |                                  ^

error: aborting due to 18 previous errors

Some errors have detailed explanations: E0436, E0658, E0797.
For more information about an error, try `rustc --explain E0436`.
